providers/azure/azure.go

```
@@ -0,0 +1,191 @@
+		Token:         newToken.String(),
+		ClusterName:   *availableCluster.Name,
+	}
+
+	leaseMap.CreateLease(newToken, leases.NewLease(*availableCluster.Name, req.ExpirationTime(time.Now())))
+	if err := k8s.SaveAnnotations(services, svc, leaseMap); err != nil {
+		log.Printf("Error saving new lease to Kubernetes annotations -- %s", err)
+		htp.Error(w, http.StatusInternalServerError, "Error saving new lease to Kubernetes annotations -- %s", err)
+		return
+	}
+
+	if err := json.NewEncoder(w).Encode(resp); err != nil {
+		log.Printf("Error encoding json -- %s", err)
+		htp.Error(w, http.StatusInternalServerError, "Error encoding json -- %s", err)
+		return
+	}
+}
+
+func getSvcsAndClusters(clusterLister ClusterLister, services k8s.ServiceGetterUpdater, k8sServiceName string) (*Map, *v1.Service, error) {
+
+	errCh := make(chan error)
+	doneCh := make(chan struct{})
+	clusterMapCh := make(chan *Map)
+	apiServiceCh := make(chan *v1.Service)
+	defer close(doneCh)
+	go func() {
+		svc, err := services.Get(k8sServiceName)
+		if err != nil {
+			select {
+			case errCh <- err:
+			case <-doneCh:
+			}
+			return
+		}
+		select {
+		case apiServiceCh <- svc:
+		case <-doneCh:
+		}
+	}()
+	go func() {
+		clusterMap, err := ParseMapFromAzure(clusterLister)
+		if err != nil {
+			select {
+			case errCh <- err:
+			case <-doneCh:
+			}
+			return
+		}
+		select {
+		case clusterMapCh <- clusterMap:
+		case <-doneCh:
+		}
+	}()
+
+	var clusterMapRet *Map
+	var apiServiceRet *v1.Service
+	for {
+		select {
+		case err := <-errCh:
+			return nil, nil, err
+		case cm := <-clusterMapCh:
+			clusterMapRet = cm
+		case svc := <-apiServiceCh:
+			apiServiceRet = svc
+		}
+		if clusterMapRet != nil && apiServiceRet != nil {
+			return clusterMapRet, apiServiceRet, nil
+		}
+	}
+}
+
+func publicKeyFile(file string) ssh.AuthMethod {
+	buffer, err := ioutil.ReadFile(file)
+	if err != nil {
+		return nil
+	}
+
+	key, err := ssh.ParsePrivateKey(buffer)
+	if err != nil {
+		return nil
```

why not return the error too?

---

doh left over code this isnt used anymore